README.md
Standartox
================

[![CRAN](https://www.r-pkg.org/badges/version/standartox)](https://CRAN.R-project.org/package=standartox)
[![Downloads](https://cranlogs.r-pkg.org/badges/standartox)](https://cran.r-project.org/package=standartox)

Standartox is a database and tool facilitating the retrieval of
ecotoxicological test data. It is based on the [EPA ECOTOX
database](https://cfpub.epa.gov/ecotox/) as well as on data from several
other chemical databases and allows users to filter and aggregate
ecotoxicological test data in an easy way. It can either be accessed via
<http://standartox.uni-landau.de> or this R-package
[standartox](https://github.com/andschar/standartox). Ecotoxicological
test data is used in environmental risk assessment to calculate effect
measures such as [TU - Toxic
Units](https://en.wikipedia.org/wiki/Toxic_unit) or [SSD - Species
Sensitivity Distributions](https://edild.github.io/ssd/) to asses
environmental toxicity of chemicals.

The project lives in two repositories:

-   [standartox-build](https://github.com/andschar/standartox-build) -
    Compiles and serves database
-   [standartox](https://github.com/andschar/standartox) - R-package

## Installation

``` r
install.packages('standartox')
# remotes::install_github('andschar/standartox') # development version
```

## Functions

Standartox consists of the two functions `stx_catalog()` and
`stx_query()`. The former allows you to retrieve a catalog of possible
parameters that can be used as an input for `stx_query()`. The latter
fetches toxicity values from the database.

### `stx_catalog()`

The function returns a list of all possible arguments that can bes use
in `stx_query()`.

``` r
require(standartox)
catal = stx_catalog()
names(catal)
```

    ##  [1] "casnr"              "cname"              "concentration_unit"
    ##  [4] "concentration_type" "chemical_role"      "chemical_class"    
    ##  [7] "taxa"               "trophic_lvl"        "habitat"           
    ## [10] "region"             "ecotox_grp"         "duration"          
    ## [13] "effect"             "endpoint"           "exposure"

``` r
catal$endpoint # access the parameter endpoint
```

| variable |      n | n_total | perc |
|:---------|-------:|--------:|-----:|
| NOEX     | 237616 |  609435 |   39 |
| LOEX     | 192718 |  609435 |   32 |
| XX50     | 179101 |  609435 |   30 |

### `stx_query()`

The function allows you to retrieve filtered and aggregated toxicity
data according to the parameters below.

| parameter          | example                                     |
|:-------------------|:--------------------------------------------|
| casnr              | 50000, 94520, 94531                         |
| cname              | 2718, 4, 3                                  |
| concentration_unit | ug/l, mg/kg, ppb                            |
| concentration_type | active ingredient, formulation, total       |
| chemical_role      | pesticide, herbicide, drug                  |
| chemical_class     | amide, aromatic, organochlorine             |
| taxa               | species, Fusarium oxysporum, Apis mellifera |
| trophic_lvl        | heterotroph, autotroph                      |
| habitat            | freshwater, terrestrial, marine             |
| region             | europe, america_north, america_south        |
| ecotox_grp         | invertebrate, plant, fish                   |
| duration           | 24, 96                                      |
| effect             | mortality, population, biochemistry         |
| endpoint           | NOEX, LOEX, XX50                            |
| exposure           | aquatic, environmental, diet                |

You can type in parameters manually or subset the object returned by
`stx_catalog()`:

``` r
require(standartox)
cas = c(Copper2Sulfate = '7758-98-7',
        Permethrin = '52645-53-1',
        Imidacloprid = '138261-41-3')
# query
l = stx_query(cas = cas,
              endpoint = 'XX50',
              taxa = grep('Oncorhynchus', catal$taxa$variable, value = TRUE), # fish genus
              exposure = 'aquatic',
              duration = c(24, 120))
```

    ## Standartox query running..
    ## Parameters:
    ## casnr: 7758-98-7, 52645-53-1, 138261-41-3
    ## duration: 24, 120
    ## endpoint: XX50
    ## exposure: aquatic
    ## taxa: Oncorhynchus clarkii, Oncorhynchus mykiss, Oncorhynchus nerka, Oncorhy...[truncated]

#### Important parameter settings

-   **CAS** (`cas =`) Can be input in the form of 7758-98-7 or 7758987
-   **Endpoints** (`endpoint =`) Only one endpoint per query is allowed:
    -   `NOEX` summarises [No observed effect
        concentration/level](https://en.wikipedia.org/wiki/No-observed-adverse-effect_level)
        (i.e. NOEC, NOEL, NOAEL, etc.)
    -   `LOEX` summarises Lowest observed effects concentration
        (i.e. LOEC, LOEL, etc.)
    -   `XX50` summarises [Half maximal effective
        concentration](https://en.wikipedia.org/wiki/EC50) (i.e. EC50,
        LC50, LD50, etc.)
-   If you leave a parameter empty Standartox will not filter for it

## Query result

Standartox returns a list object with five entries.

-   `l$filtred` and `l$filtered_all` contain the filtered Standartox
    data set (the former only is a shorter and more concise version of
    the latter):

| cas       | cname          | concentration | concentration_unit | effect    | endpoint |
|:----------|:---------------|--------------:|:-------------------|:----------|:---------|
| 7758-98-7 | cupric sulfate |        1100.0 | ug/l               | mortality | XX50     |
| 7758-98-7 | cupric sulfate |          18.9 | ug/l               | mortality | XX50     |
| 7758-98-7 | cupric sulfate |          46.4 | ug/l               | mortality | XX50     |

-   `l$aggregated` contains the several aggregates of the Standartox
    data:

    -   `cname`, `cas` - chemical identifiers
    -   `min` - Minimum
    -   `tax_min` - Most sensitive taxon
    -   **`gmn`** - **Geometric mean**
    -   `amn` - Arithmetic mean
    -   `sd` - Standard Deviation of the arithmetic mean
    -   `max` - Maximum
    -   `tax_max` - Most insensitive taxon
    -   `n` - Number of distinct taxa used for the aggregation
    -   `tax_all` - Concatenated string of all taxa used for the
        aggregation

| cname          | cas         |          min | tax_min              |          gmn |         max |
|:---------------|:------------|-------------:|:---------------------|-------------:|------------:|
| cupric sulfate | 7758-98-7   | 6.813740e+01 | Oncorhynchus clarkii | 1.330055e+02 |    263.6153 |
| imidacloprid   | 138261-41-3 | 2.291000e+05 | Oncorhynchus mykiss  | 2.291000e+05 | 229100.0000 |
| permethrin     | 52645-53-1  | 1.896481e+00 | Oncorhynchus gilae   | 4.505877e+00 |     17.0000 |

-   `l$meta` contains meta information on the request:

| variable           | value               |
|:-------------------|:--------------------|
| accessed           | 2022-08-06 20:59:07 |
| standartox_version | 20210315            |

## Example: *Oncorhynchus*

Let’s say, we want to retrieve the 20 most tested chemicals on the genus
*[Oncorhynchus](https://en.wikipedia.org/wiki/Oncorhynchus)*. We allow
for test durations between 48 and 120 hours and want the tests
restricted to active ingredients only. Since we are only interested in
the half maximal effective concentration, we choose XX50 as our
endpoint. As an aggregation method we choose the geometric mean.

``` r
require(standartox)
l2 = stx_query(concentration_type = 'active ingredient',
               endpoint = 'XX50',
               taxa = grep('Oncorhynchus', catal$taxa$variable, value = TRUE), # fish genus
               duration = c(48, 120))
```

    ## Standartox query running..
    ## Parameters:
    ## concentration_type: active ingredient
    ## duration: 48, 120
    ## endpoint: XX50
    ## taxa: Oncorhynchus clarkii, Oncorhynchus mykiss, Oncorhynchus nerka, Oncorhy...[truncated]

We subset the retrieved data to the 20 most tested chemicals and plot
the result.

``` r
require(data.table)
dat = merge(l2$filtered, l2$aggregated, by = c('cas', 'cname'))
cas20 = l2$aggregated[ order(-n), cas ][1:20]
dat = dat[ cas %in% cas20 ]
```

``` r
require(ggplot2)
ggplot(dat, aes(y = reorder(cname, -gmn))) +
  geom_point(aes(x = concentration, col = 'All values'),
             pch = 1, alpha = 0.3) +
  geom_point(aes(x = gmn, col = 'Standartox value\n(Geometric mean)'),
             size = 3) +
  scale_x_log10(breaks = c(0.01, 0.1, 1, 10, 100, 1000, 10000),
                labels = c(0.01, 0.1, 1, 10, 100, 1000, 10000)) +
  scale_color_viridis_d(name = '') +
  labs(title = 'Oncorhynchus EC50 values',
       subtitle = '20 most tested chemicals',
       x = 'Concentration (ppb)') +
  theme_minimal() +
  theme(axis.title.y = element_blank())
```

![](README_files/figure-gfm/unnamed-chunk-12-1.png)<!-- -->

## Usage

We ask you to use the API service thoughtfully, which means to run the
`stx_query()` only once and to re-run it only when parameters change or
you want to query new versions. Here is an example of how to easily
store the queried data locally from within R.

``` r
run = FALSE # set to TRUE for the first run
if (run) {
  l2 = stx_query(concentration_type = 'active ingredient',
                 endpoint = 'XX50',
                 taxa = grep('Oncorhynchus', catal$taxa$variable, value = TRUE), # fish genus
                 duration = c(48, 120))
  saveRDS(l2, file.path('path/to/directory', 'data.rds'))
  
} else {
  l2 = readRDS(file.path('path/to/directory', 'data.rds'))
}

# put rest of the script here
# ...
```

## Article

The article on Standartox is published
[here](https://www.mdpi.com/2306-5729/5/2/46).

## Information

### Contributors

-   [Andreas Scharmüller](https://andschar.github.io)

### Want to contribute?

Check out our [contribution guide
here](https://github.com/andschar/standartox/blob/master/CONTRIBUTING.md).

### Meta

-   Please report any [issues, bugs or feature
    requests](https://github.com/andschar/standartox/issues)
-   License: MIT
-   Get citation information for the standartox package in R doing
    `citation(package = 'standartox')`


NEWS
# standartox 0.0.2

* Fix CRAN NOTE & Policy 

# standartox 0.0.1

* Initial release


R/standartox.R
#' Retrieve data catalog
#' 
#' Retrieve a data catalog for all variables (and their values) that can be retrieved with stx_query()
#' 
#' @param vers integer; Choose the version of the EPA Ecotox on which Standartox is based on. NULL (default) accesses the most recent version
#' 
#' @return Returns a list of data.frames containing information on data base variables
#' @author Andreas Scharmueller \email{andschar@@protonmail.com}
#' @examples
#' \donttest{
#' # might fail if API is not available
#' stx_catalog()
#' }
#' 
#' @export
stx_catalog = function(vers = NULL) {
  # request
  body = list(vers = vers)
  # POST
  message('Retrieving Standartox catalog..')
  res = try(httr::POST(
    url = file.path(stx_domain(), 'catalog'),
    body = body,
    encode = 'json',
  ), silent = TRUE)
  if (inherits(res, 'try-error') ||
      ! httr::status_code(res) %in% c(200, 400)) {
    stx_availability(res)
    
    return(NULL)
  }
  cont = httr::content(res, type = 'text', encoding = 'UTF-8')
  
  jsonlite::fromJSON(cont)
}

#' Retrieve Standartox toxicity values
#'
#' Retrieve toxicity values from the Standartox data base \url{http://standartox.uni-landau.de/}.
#'
#' @import httr jsonlite fst data.table
#'
#' @param cas character, integer; Limit data base query to specific CAS numbers, multiple entries possible (e.g. 1071-83-6, 1071836), NULL (default).
#' @param concentration_unit character; Limit data base query to specific concentration units (e.g. ug/l - default).
#' @param concentration_type character; Limit data base query to specific concentration types, can be one of NULL (default), 'active ingredient', 'formulation', 'total', 'not reported', 'unionized', 'dissolved', 'labile'. See \url{https://cfpub.epa.gov/ecotox/pdf/codeappendix.pdf} p.4.
#' @param duration integer vector of length two; Limit data base query to specific test durations (hours) (e.g. c(24, 48)). NULL (default).
#' @param endpoint character; Choose endypoint type, must be one of 'XX50' (default), 'NOEX', 'LOEX'.
#' @param effect character; Limit data base query to specific effect groups, multiple entries possible (e.g. 'Mortality', 'Intoxication', 'Growth'). See \url{https://cfpub.epa.gov/ecotox/pdf/codeappendix.pdf} p.95. NULL (default).
#' @param exposure character; Choose exposure type, (e.g. aquatic, environmental, diet). NULL (default).
#' @param chemical_role character; Limit data base query to specific chemical roles (e.g. insecticide), multiple entries possible. NULL (default).
#' @param chemical_class character; Limit data base query to specific chemical classes (e.g. neonicotinoid), multiple entries possible. NULL (default).
#' @param taxa character; Limit data base query to specific taxa, multiple entries possible. NULL (default).
#' @param ecotox_grp character; Convenience grouping of organisms in ecotoxicology, must be one of NULL (default), 'invertebrate', 'fish', 'plant_land', 'macrophyte', 'algae'.
#' @param trophic_lvl character; Trophic level of organism, must be one of NULL (default), 'autotroph', 'heterotroph'.
#' @param habitat character; Limit data base query to specific organism habitats, can be one of NULL (default) 'marine', 'brackish', 'freshwater', 'terrestrial'.
#' @param region character; Limit data base query to organisms occurring in specific regions, can be one of NULL (default) 'africa', 'america_north', 'america_south', 'asia', 'europe', 'oceania'.
#' @param vers integer; Choose the version of the EPA Ecotox on which Standartox is based on. NULL (default) accesses the most recent version.
#' @param ... currently not used
#'
#' @return Returns a list of three data.tables (filtered data base query results, aggregated data base query results, meta information)
#' @author Andreas Scharmueller \email{andschar@@protonmail.com}
#' 
#' @examples
#' \donttest{
#' # might fail if API is not available
#' stx_query('1071-83-6')
#' stx_query(cas = '1071-83-6',
#'           duration = c(48, 120),
#'           concentration_unit = 'ug/l')
#' stx_query(cas = '1071-83-6',
#'           duration = c(48, 120),
#'           concentration_unit = 'ug/l',
#'           endpoint = 'XX50')
#' }
#' 
#' @export
#'
stx_query = function(cas = NULL,
                     concentration_unit = NULL,
                     concentration_type = NULL,
                     duration = NULL,
                     endpoint = c('XX50', 'NOEX', 'LOEX'),
                     effect = NULL,
                     exposure = NULL,
                     chemical_role = NULL,
                     chemical_class = NULL,
                     taxa = NULL,
                     ecotox_grp = NULL,
                     trophic_lvl = NULL,
                     habitat = NULL,
                     region = NULL,
                     vers = NULL,
                     ...) {
  # to avoid NOTE in R CMD check --as-cran
  casnr = outlier = concentration = cname = NULL
  # DEBUG
  # browser()
  # cas = '1071-83-6'; concentration_unit = NULL; concentration_type = NULL; duration = NULL; endpoint = 'XX50'; effect = NULL; exposure = NULL; chemical_role = NULL; chemical_class = NULL; taxa = NULL; ecotox_grp = NULL; trophic_lvl = NULL; habitat = NULL; region = NULL; vers = NULL
  # checks
  endpoint = match.arg(endpoint)
  # request
  body = list(casnr = cas,
              concentration_unit = concentration_unit,
              concentration_type = concentration_type,
              duration = duration,
              endpoint = endpoint,
              effect = effect,
              exposure = exposure,
              chemical_role = chemical_role,
              chemical_class = chemical_class,
              taxa = taxa,
              ecotox_grp = ecotox_grp,
              trophic_lvl = trophic_lvl,
              habitat = habitat,
              region = region,
              vers = vers)
  # POST
  stx_message(body)
  res = try(httr::POST(
    file.path(stx_domain(), 'filter'),
    body = body,
    encode = 'json',
  ), silent = TRUE)
  if (inherits(res, 'try-error') ||
      ! httr::status_code(res) %in% c(200, 400)) {
    stx_availability(res)
    
    return(NULL)
  }
  if (res$status_code == 400) {
    warning(jsonlite::fromJSON(httr::content(res, type = 'text', encoding = 'UTF-8')))
    out_fil = data.table(NA)
    filtered = data.table(NA)
    out_agg = data.table(NA)
  }
  if (res$status_code == 200) {
    out_fil = read_bin_vec(res$content, type = 'fst')
    if (nrow(out_fil) == 0) {
      warning('No results found.')
      out_fil = data.table(NA)
      filtered = data.table(NA)
      out_agg = data.table(NA)
    } else {
      # CAS column (not sent through API to reduce size)
      out_fil[ , cas := cas_conv(casnr) ][ , casnr := NULL ]
      # outliers
      para = c('cas',
               'concentration_unit',
               'concentration_type',
               'duration',
               'tax_taxon',
               'effect',
               'endpoint',
               'exposure')
      out_fil[ ,
               outlier := flag_outliers(concentration),
               by = para ]
      # colorder
      nam = names(out_fil)
      col_order = c('cname', 'cas', 'iupac_name', 'inchikey', 'inchi', 'molecularweight',
                    'result_id', 'endpoint', 'effect', 'exposure', 'trophic_lvl', 'ecotox_grp', 'concentration_type',
                    'concentration', 'concentration_unit', 'concentration_orig', 'concentration_unit_orig',
                    'duration', 'duration_unit', 'outlier',
                    c('species_number', grep('tax_|hab_|reg_', nam, value = TRUE)),
                    grep('cro_|ccl_', nam, value = TRUE),
                    grep('ref', nam, value = TRUE))
      setcolorder(out_fil, col_order)
      # order
      setorder(out_fil, cname)
      # id
      col_id = c('cname', 'cas', 'inchikey', 'inchi', 'result_id', 'species_number', 'ref_number')
      id = out_fil[ , .SD, .SDcols = col_id ]
      # short
      col_short = c('cname', 'cas', 'inchikey',
                    'endpoint', 'effect', 'exposure', 'trophic_lvl', 'ecotox_grp', 'concentration_type',
                    'concentration', 'concentration_unit', 'concentration_orig', 'concentration_unit_orig',
                    'duration', 'duration_unit', 'outlier',
                    grep('tax_', nam, value = TRUE))
      filtered = out_fil[ , .SD, .SDcols = col_short ]
      # aggregate
      out_agg = suppressWarnings(
        stx_aggregate(out_fil)
      )
    }
  }
  # meta
  out_meta = stx_meta()
  # return
  list(filtered = filtered,
       filtered_all = rm_col_na(out_fil),
       # TODO id = id,
       aggregated = out_agg,
       meta = out_meta)
}

#' Retrieve chemical data
#' 
#' Retrieve data on all chemicals in Standartox.
#' 
#' @return Returns a data.table containing informaiton on chemicals in Standartox.
#' @author Andreas Scharmueller \email{andschar@@protonmail.com}
#' @examples
#' \donttest{
#' # might fail if API is not available
#' stx_chem()
#' }
#' 
#' @export
#' 
stx_chem = function() {
  message('This endpoint is still experimental.')
  res = try(httr::GET(
    url = file.path(stx_domain(), 'chem'),
    encode = 'json'
  ), silent = TRUE)
  if (inherits(res, 'try-error') ||
      ! httr::status_code(res) %in% c(200, 400)) {
    stx_availability(res)
    
    return(NULL)
  }
  out = read_bin_vec(res$content, type = 'fst')
  setnames(out, sub('ccl_|cro_', '', names(out)))
  out
}

#' Retrieve taxa data
#' 
#' Retrieve data on all taxa in Standartox.
#' 
#' @return Returns a data.table containing informaiton on taxa in Standartox.
#' @author Andreas Scharmueller \email{andschar@@protonmail.com}
#' @examples
#' \donttest{
#' # might fail if API is not available
#' stx_taxa()
#' }
#' 
#' @export
#' 
stx_taxa = function() {
  message('This endpoint is still experimental.')
  res = try(httr::GET(
    url = file.path(stx_domain(), 'taxa'),
    encode = 'json'
  ), silent = TRUE)
  if (inherits(res, 'try-error') ||
      ! httr::status_code(res) %in% c(200, 400)) {
    stx_availability(res)
    
    return(NULL)
  }
  out = read_bin_vec(res$content, type = 'fst')
  setnames(out, sub('hab_|reg_', '', names(out)))
  out
}

#' Retrieve meta data
#' 
#' @author Andreas Scharmüller
#' @noRd
#' 
stx_meta = function(vers = NULL) {
  # request
  body = list(vers = vers)
  # POST
  res = try(httr::POST(
    url = file.path(stx_domain(), 'meta'),
    body = body,
    encode = 'json'
  ), silent = TRUE)
  if (inherits(res, 'try-error') ||
      ! httr::status_code(res) %in% c(200, 400)) {
    stx_availability(res)
    
    return(NULL)
  }
  cont = httr::content(res, type = 'text', encoding = 'UTF-8')
  out = jsonlite::fromJSON(cont)
  out
}

# IDEA
# microbenchmark::microbenchmark({
# col_test = c('result_id', 'cas', 'species_number', 'ref_number',
#              'concentration', 'concentration_unit', 'concentration_orig', 'concentration_unit_orig',
#              'concentration_type', 'duration', 'duration_unit',
#              'endpoint', 'effect', 'exposure', 'outlier')
# col_chem = c('cname', 'iupac_name', 'cas', 'cas', 'inchikey', 'inchi',
#              'molecularweight',
#              grep('cro|ccl', names(out_fil), value = TRUE))
# col_taxa = c(
#   grep('tax|hab|reg', names(out_fil), value = TRUE))
# col_refs = grep('ref', names(out_fil), value = TRUE)
# l = list(
#   test = unique(rm_col_na(out_fil[ , .SD, .SDcols = col_test ])),
#   chem = unique(rm_col_na(out_fil[ , .SD, .SDcols = col_chem ])),
#   taxa = unique(rm_col_na(out_fil[ , .SD, .SDcols = col_taxa ])),
#   refs = unique(rm_col_na(out_fil[ , .SD, .SDcols = col_refs ])),
#   aggregated = out_agg,
#   meta = stx_meta()
# )
# })
# 


#' Function to aggregate filtered test results
#' 
#' 
#' @author Andreas Scharmueller \email{andschar@@protonmail.com}
#' @noRd
#'  
stx_aggregate = function(dat = NULL) {
  # assign variables to avoid R CMD check NOTES
  . = concentration = cname = cas = tax_taxon = gmn = gmnsd = n = NULL
  # checking
  if (is.null(dat)) stop('Provide table.')
  # aggregation
  dat[
    ,
    .(gmn = gm_mean(concentration),
      gmnsd = gm_sd(concentration),
      n = .N),
    .(cname, cas, tax_taxon) 
    ][
      ,
      .(min = min(gmn),
        tax_min = .SD[ which.min(gmn), tax_taxon ],
        gmn = gm_mean(gmn),
        gmnsd = gm_sd(gmnsd),
        max = max(gmn),
        tax_max = .SD[ which.max(gmn), tax_taxon ],
        n = sum(n),
        tax_all = paste0(sort(unique(tax_taxon)), collapse = ', ')),
      .(cname, cas)
  ]
}

#' Connection URL
#' 
#' @author Andreas Scharmüller
#' @noRd
#' 
stx_domain = function() {
  baseurl = 'http://139.14.20.252'
  # baseurl = 'http://127.0.0.1' # DEBUG
  port = 8000
  dmn = paste0(baseurl, ':', port)
  
  dmn
}

#' Check availability of connection
#' 
#' @author Andreas Scharmüller
#' @noRd
#' 
stx_availability = function(res) {
    msg = paste0(
    'The standartox web service seems currently unavailable.
    Please try again after some time. Should it still not work then, please file an issue here:
https://github.com/andschar/standartox/issues
Error code: ',res)
    message(msg)
}
